C++ cheatsheet (#282)
ferrous-systems/rust-training: training-slides/src/cpp-cheatsheet.md

    * Value Semantics & Data-oriented programming over complex object graphs
* Writing Safe Rust gets you the following benefits instantly:
    * References are checked to be valid while in use
    * running UBSAN, ASAN, THREADSAN and RAII analysis at compile time, without the performance penalty at runtime
    * all code you depend on is also analyzed under the same constraints by the compiler

## Installation

* Cargo is the package manager, not the compiler
    * Like e.g. CMake, Cargo manages the compiler (rustc) and linker for you
    * Uses the system linker
    * Cargo does not have a separate "configure" stage like CMake
    * It's possible to use rustc directly, but very rarely needed
* rustc is the compiler
    * LLVM is the default codegen backend
    * [experimental gcc backend](https://rust-gcc.github.io) has not yet stabilized
* [CodeLLDB](https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb) is a reliable debugger setup
* Rust comes with an auto-formatter (rustfmt) that should be used by default
    * More consistent & reliable than e.g. clang-format
    * Most code in the Rust ecosystem uses rustfmt style

## Basic Types

### Integers

> * No `int` type, use `i32` instead

Reviewer: Are there any platforms left in production where a C++ `int` is *not* going to match a Rust `i32`?

Author: None that I'm aware of.
The point this is trying to make is just to emphasize to C++ devs that that there simply is no type named "int" (even if it's just an alias for i32).

In C++, there are explicitly sized types (e.g. `std::int32_t`), but `int` is still what everyone uses.
So it can be strange to use the explicitly sized version all of a sudden, just because people are so used to writing `int`.

Is there a better way to phrase this?

Reviewer: I tried a bunch of alternative phrasings and I didn't like them in the end so ... no ... let's got with what you put!

* Use `usize` wherever you would use `std::size_t` in C++
* Integers in Rust cannot be used as booleans
    * Use explicit `if my_number != 0` instead of `if (my_number)`

### Strings and arrays

> * `char` in Rust represents an actual Unicode character (32-bit)

Reviewer: I would say it represents a Unicode Scalar Value, and they are 21-bit values.

Author: Done

```

## Compound Types

### Structs

No `class` type in Rust
* Use `struct` instead
* Only data members are declared inside the struct
* Member functions are declared outside the struct itself
* No inheritance, Rust uses composition and traits instead (foreshadowing 👻)

### Construction

Construction in Rust is similar to [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization) with designated initializers.

> There are no constructors in Rust, use "static" member functions instead to uphold invariants before construction.

Reviewer: ```suggestion
There are no constructors in Rust, use "static" member functions (Rust calls them 'associated functions') instead to uphold invariants before construction.
```

Author: Done

### Enums

A Rust `enum` is most similar to a `std::variant`, it can hold data in each of the variants.

* By default, the compiler chooses an optimal layout
    * Representation can be chosen explicitly with an attribute, e.g. `#[repr(u8)]`
    * Enum values can only be of the declared variants - not any integer

>     * Cannot use enums as flags directly - use libraries like bitflags

Reviewer: ```suggestion
    * Cannot use enums as flags directly - use libraries like [`bitflags`](https://crates.io/crates/bitflags)
```

1. Moving the value
2. Copying the data into a new value
3. Handing out a reference to the value

The difference is in the defaults:

Taking something by-value in Rust always means **moving** the value, not copying.
In comparison, copying is explicit with `.clone()` and references are explicit with `&` and `&mut`.

The second important difference is that a move does not leave an object behind, when moving out of an object, the moved-from object is no longer accessible.

### References Cheat Sheet

|                                 | C++                      | Rust                               |
|---------------------------------|--------------------------|------------------------------------|
| Shared Reference Declaration    | `const std::string &arg` | `arg: &String`<br/>or: `arg: &str` |
| Shared Reference Passing        | `foo(arg);`              | `foo(&arg);`                       |
| Exclusive Reference Declaration | `std::string &arg`       | `arg: &mut String`                 |
| Exclusive Reference Passing     | `foo(arg);`              | `foo(&mut arg);`                   |

### Rules around References/Borrowing

> Rusts references are similar to C++ references, but many rules/best-practices that C++ holds you responsible for are enforced at compile time:

Reviewer: ```suggestion
Rust's references are similar to C++ references, but many rules/best-practices that C++ holds you responsible for are enforced at compile time:
```

Author: Done

* No inheritance, Rust uses composition and traits instead (foreshadowing 👻)

### Construction

Construction in Rust is similar to [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization) with designated initializers.

There are no constructors in Rust, use "static" member functions (Rust calls them "associated functions") instead to uphold invariants before construction.

### Enums

A Rust `enum` is most similar to a `std::variant`, it can hold data in each of the variants.

* By default, the compiler chooses an optimal layout
    * Representation can be chosen explicitly with an attribute, e.g. `#[repr(u8)]`
    * Enum values can only be of the declared variants - not any integer
    * Cannot use enums as flags directly - use libraries like [`bitflags`](https://crates.io/crates/bitflags)

## Ownership and Borrowing

Like C++, Rust fundamentally has three ways to pass ownership around:

1. Moving the value
2. Copying the data into a new value
3. Handing out a reference to the value

The difference is in the defaults:

Taking something by-value in Rust always means **moving** the value, not copying.

> In comparison, copying is explicit with `.clone()` and references are explicit with `&` and `&mut`.

Reviewer 2: I would be a bit more precise with wording here - since you can have copying without explicit `.clone()` calls of `Copy`able objects by using ` a = b`, [see the stdlib](https://doc.rust-lang.org/std/marker/trait.Copy.html#whats-the-difference-between-copy-and-clone).

Their example of a plain struct is neat, but where it can be surprising for C++ folks would probably be with [stack-allocated arrays](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=00d1ff756461161747fa641fa7cc1d75):

```rust
fn main() {
    let arr1 = [1; 10_000];
    let arr2 = arr1;
    println!("{}", arr2.len());
}
```

Each instance of `foo = arr1` would memcopy 10,000 elements, which would probably be unexpected!

Author: I added a section on `Copy` specifically to put it into context for C++ devs.

| `std::span<T>`            | `&mut [T]`                        |                                          |
| `std::vector<T>`          | `Vec<T>`                          |                                          |
| `std::string_view`        | `&str`                            | Rust: UTF-8                              |
| `const char *`            | `&'static str`                    | String literals only<br/> Rust: UTF-8    |
| `std::string`             | `String`                          | Rust: UTF-8                              |
| `std::deque`              | `VecDeque`                        | Best match, slightly different internals |
| `std::unordered_map<K,V>` | `std::collections::HashMap<K,V>`  |                                          |
| `std::map<K,V>`           | `std::collections::BTreeMap<K,V>` | Best match, slightly different internals |
| `std::unordered_set<T>`   | `std::collections::HashSet<T>`    |                                          |
| `std::set<T>`             | `std::collections::BTreeSet<T>`   | Best match, slightly different internals |

## Iterators

Iterators in Rust are self-contained. No need for an end iterator.

> Most algorithms are implemented directly on Iterator trait, not as separate functions.

Reviewer 2: ```suggestion
Most algorithms are implemented directly on the Iterator trait, not as separate functions.
```

So this C++ code:

```cpp
auto numbers = std::vector{ 1, 2, 3 };

auto odd = std::find_if(
                numbers.begin(),
                numbers.end(),
                [](const auto& number) { return number % 2 == 0; });

if (odd != numbers.end()) {
    std::cout << *odd << std::endl;
}
```

becomes:

```rust
let numbers = vec![1, 2, 3];

let odd = numbers
            .iter()
            .find(|number| *number % 2 == 0);

if let Some(odd) = odd {
    println!("{odd}");
}
```

## Imports and Modules

Rust modules work like C++ modules, not like `#include` files.
They export certain symbols (i.e. types/functions) under a given name.

> They and are not included as in-place text.

Reviewer 2: ```suggestion
They are not included as in-place text.
```

Author: Done

Modules are also important for scoping in Rust.
Unlike namespaces, modules and all items in them have their own visibility (e.g. `pub` or not).

Inside a module, all items can access each other, but items are only accessible from outside the module if they are `pub`.
You can think of everything inside the same module as being a `friend` (as in the C++ keyword `friend`) of everything else inside the same module.

e.g. this works in Rust:

```rust
mod config {
    // private struct inside the module
    struct Config {
        // with private members
        color_enabled: bool,
        unicode_supported: bool,
    }

    // non-member function can still access the members inside the same module
    fn is_color_enabled(config: &Config) -> bool {
        config.color_enabled
    }
}

// But this would not work, as it's outside of `mod config`:
// fn is_unicode_supported(config: &Config) -> bool {
//     config.is_unicode_supported
// }
```

## Good Design Practices

### Operator Overloading with Traits

> Traits like `PartialEq`, `PartialOrd`, etc. are Rusts way of operator overloading.

Reviewer 2: ```suggestion
Traits like `PartialEq`, `PartialOrd`, etc. are Rust´s way of operator overloading.
```

Author: Done

> If a type implements the right trait, the operator (e.g. `==`, `<`, etc.) are available for the type.

Reviewer 2: ```suggestion
If a type implements the right trait, the corresponding operator (e.g. `==`, `<`, etc.) is available for the type.
```

Author: Done

### Construction

Construction in Rust is similar to [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization) with designated initializers.

There are no constructors in Rust, use "static" member functions (Rust calls them "associated functions") instead to uphold invariants before construction.

### Enums

A Rust `enum` is most similar to a `std::variant`, it can hold data in each of the variants.

* By default, the compiler chooses an optimal layout
    * Representation can be chosen explicitly with an attribute, e.g. `#[repr(u8)]`
    * Enum values can only be of the declared variants - not any integer
    * Cannot use enums as flags directly - use libraries like [`bitflags`](https://crates.io/crates/bitflags)

## Ownership and Borrowing

Like C++, Rust fundamentally has three ways to pass ownership around:

1. Moving the value
2. Copying the data into a new value
3. Handing out a reference to the value

The difference is in the defaults:

Taking something by-value in Rust by default means **moving** the value, not copying.
In comparison, copying is usually explicit with `.clone()` and references are explicit with `&` and `&mut`.

> The second important difference is that a move does not leave an object behind, when moving out of an object, the moved-from object is no longer accessible.

Author: Done

Lifetimes are here to help!
They give Rust a way to express these relationships between references and whatever they reference as a language construct.
So instead of documenting these important relations as text and hoping people read the documentation, lifetimes allow Rust to **enforce them at compile time!**

You will probably recognize many of the issues that the Rust compiler prevents as common pitfalls in C/C++.

## Cargo Workspaces

In Rust, the smallest compilation unit is a whole crate, not just one source file.

At the time of writing (05/2025), the Rust compiler is largely single-threaded.
Adding more crates to your build therefore tends to *speed up* compilation by allowing Cargo to compile multiple compilation units in parallel.

In general, splitting your project into multiple crates is very common and good practice in Rust.
Package management is considerably easier compared to C++, do not be afraid to split your project into multiple smaller packages.

## Heap Allocation (Box and Rc)

Heap allocation in Rust is almost always done via smart pointers.

Rust smart pointers are very similar to the C++ equivalents, but cannot be null!

> If you need a nullable smart pointer use them together with `Option<T>`, e.g. `Box<Option<T>>`.

Author: Absolutely fair point, done.

## Thread Safety (Send/Sync, Arc, Mutex)

### Mutexes

In C++, you have to manually ensure that you lock the right mutex for the right data.
A Rust `Mutex<T>` can be thought of as a `std::mutex` together with some arbitrary data of type `T`.
The data `T` is owned and protected by the mutex.

`Mutex::lock` returns a `MutexGuard`, which is similar to a `std::lock_guard`.
Note that it does **not** include a deadlock avoidance algorithm like `std::scoped_lock`!

`MutexGuard` dereferences (mutably) to the inner `T` data, thereby granting mutable access.

### Arc vs. shared_ptr

Arc is Rust's `std::shared_ptr` and also uses atomics for thread-safe reference counting.

In C++ you must take care to never mutate a `shared_ptr` from multiple threads, as the `shared_ptr` itself is not thread-safe, only the internal reference counting.
`Arc` is implemented the same way, but as it is impossible to gain two mutable references to the same `Arc` in Rust, you do not need to worry about this.

### Atomics

Atomics in Rust are pretty much the same as in C++.

> However, they are generic types, but multiple concrete types.

Author: Done

### Capturing data in closures

Closures are the Rust equivalent of C++ lambdas.
Unlike C++, Rust does not have an explicit capture list.
By default, every outside variable is captured by reference.

Therefore `|arg| { ... }` is the Rust equivalent of `[&](auto arg) { ... }`.
To capture everything by-value (e.g. by move), add the `move` keyword (e.g. `move || { ... }`).

If you need to specify explicitly which types to capture by-value/by-copy/by-reference, use `move` together with an added scope.

e.g. this C++ capture list:
```cpp
auto by_move = std::string("Hello Move");
auto by_copy = std::string("Hello Copy");
auto by_reference = std::string("Hello Reference");

auto lambda = [by_copy, by_move=move(by_move), &by_reference]() {
    // ...
};
```

becomes:

```rust
let by_value = "Hello Move".to_owned();
let by_copy = "Hello Copy".to_owned();
let by_reference = "Hello Reference".to_owned();

// create a scope for the closure
let closure = {
    let by_reference = &by_reference; // shadow with a reference
    // let by_reference = &mut by_reference; // or mutable reference
    let by_copy = by_copy.clone(); // or explicit clone

>     // The rest is captured by move

Reviewer: Technically, everything is moved, but some of the variables you are moving are now 'reference-to-thing', so 'thing' isn't moved, only the reference.

Reviewer: Can you push the changes?
